### PR TITLE
fix ddpo finetune issue in torch2.5.1

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by sywangyi** [comment]
loss is nan when "sdp_on_bf16" is not set in torch 2.5.1

**Comment by sywangyi** [comment]
python ddpo.py \
  --num_epochs=200 \
  --train_gradient_accumulation_steps=1 \
  --sample_num_steps=50 \
  --sample_batch_size=6 \
  --train_batch_size=3 \
  --sample_num_batches_per_epoch=4 \
  --per_prompt_stat_tracking=True \
  --per_prompt_stat_tracking_buffer_size=32 \
  --train_learning_rate=1e-05 \
  --tracker_project_name="stable_diffusion_training" \
  --log_with="tensorboard" \
  --use_habana \
  --use_hpu_graphs \
  --bf16 \
  --hf_hub_model_id="ddpo-finetuned-stable-diffusion" \
  --push_to_hub False

inference
```

import torch

from optimum.habana import GaudiConfig
from optimum.habana.trl import GaudiDefaultDDPOStableDiffusionPipeline

gaudi_config = GaudiConfig.from_pretrained("Habana/stable-diffusion")
model_id = "CompVis/stable-diffusion-v1-4"
lora_model_id = "ddpo-finetuned-stable-diffusion"
pipeline = GaudiDefaultDDPOStableDiffusionPipeline(
    model_id,
    use_habana=True,
    use_hpu_graphs=True,
    gaudi_config=gaudi_config,
)
pipeline.sd_pipeline.load_lora_weights(lora_model_id)
device = torch.device("hpu")

# memory optimization
pipeline.vae.to(device, torch.bfloat16)
pipeline.text_encoder.to(device, torch.bfloat16)
pipeline.unet.to(device, torch.bfloat16)

prompts = ["lion", "squirrel", "crab", "starfish", "whale", "sponge", "plankton"]
results = pipeline(prompts)

for prompt, image in zip(prompts, results.images):
    image.save(f"{prompt}.png")

```

no error but seems like the output PNG has header information only.
Size is 842 bytes always.

**Comment by sywangyi** [comment]
training loss is NAN during finetune......, not see it in torch2.4

**Comment by sywangyi** [comment]
someone report the bug to me, it's about finetune ddpo model, then, use it for inference. black image output. it's found in 1.19 not 1.18. the 1.19 is released with torch 2.5.1. 
In rls note. we see
Starting from PyTorch 2.5.1,
torch.nn.functional.scaled_dot_product_attention automatically casts inputs to FP32, which can lead to performance degradation or GC failures. To avoid this, set torch._C._set_math_sdp_allow_fp16_bf16_reduction(True) to enable torch.nn.functional.scaled_dot_product_attention to operate on BF16. Without this setting, the inputs will always be cast to FP32.

I found NAN in loss is also caused by the ops. and  torch._C._set_math_sdp_allow_fp16_bf16_reduction(True) could restore the original result in 1.19

**Review by jiminha — APPROVED** [review]

**Review by regisss — APPROVED** [review]
